### DISCUSSION TITLE
Can I use SVG images with percentage width and height?

### DISCUSSION THREAD

**Discussion by i-do** [datenstrom/community#407] [Ask a question] [answered]
Find that svg files doesn't scale at all. 
I'm always see the 100% in spite of parameters.

Found [some css tricks](https://css-tricks.com/scale-svg/) to make it works but do not know how to make it on the code.

**Comment by markseuffert** [comment]
I can investigate as soon as there is detailed information to reproduce a problem.

**Comment by GiovanniSalmeri** [comment]
I could reproduce the problem of @i-do with with an [image generated by Inkscape](https://mondodomani.org/temp/clock2.svg) and the following shortcuts:

`[image clock.svg Example - 100 100]`

The result is:

`Something went wrong. Can't write file 'media/thumbnails/clock-100x100.svg'!`

With the shortcut

`[image clock.svg Example - 30%]`

The result is:

`<img src="/media/images/clock.svg" alt="Example" title="Example" />`
(Percent is ignored.)

After looking carefully at the code, I think I have found the problem in the regex used to extract the svg tag in `core.php`. After rewriting it as `"/<svg\s(.*?)>/s"` all seems to work well. I have also tried to make more robust the other regex's, to add a fallback in case width and height are lacking, and in case also viewBox is lacking. The resulting code is as follows:

```
                if (!feof($fileHandle) && preg_match("/<svg\s.*?>/s", $dataBuffer, $matches)) {
                    if (preg_match("/\swidth=[\"'](\d+)(?:px)?[\"']/s", $matches[0], $tokens)) $width = $tokens[1];
                    if (preg_match("/\sheight=[\"'](\d+)(?:px)?[\"']/s", $matches[0], $tokens)) $height = $tokens[1];
                    if (!$width || !$height) {
                        if (preg_match("/\sviewBox=[\"'](\d+)[,\s]\s*(\d+)[,\s]\s*(\d+)[,\s]\s*(\d+)[\"']/s", $matches[1], $tokens)) {
                            $width = ceil($tokens[2]-$tokens[0]);
                            $height = ceil($tokens[3]-$tokens[1]);
                        } else {
                            $width = 300;
                            $height = 150;
                        }
                    }
                    $type = $fileType;
```

(I have written also a version with `simplexml_load_string`, but I guess it is overkill.) Anyway this code deals only with the commonest case of width and height expressed in px: they could be expressed in any unit or in a perecentage of viewBox. But perhaps this is enough. @markseu Do you think it is OK? I didn't write a pull request because I do not know if I should make it in yellow or in yellow-extensions. Thank you!

**Comment by markseuffert** [comment]
Can you give more details, @i-do.  Please give some real life SVG files and which shortcuts you used.

**Comment by i-do** [comment]
[Mark](@markseu), I use this two svg-files:
1. [4a0-favicon.svg](https://kok59.ru/files/storage/4a0-favicon.svg)
Size is 256 × 256 px
`<svg height="256" viewBox="0 0 256 256" width="256" xmlns="http://www.w3.org/2000/svg">`

With parameters [image 4a0-favicon.svg - 300 300] and it perfectly works. 

2. [4a0-favicon-A.svg](https://kok59.ru/files/storage/4a0-favicon-A.svg)
Size is 100%
`<svg width="100%" height="100%" viewBox="0 0 445 229" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" xmlns:serif="http://www.serif.com/" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:1.41421;">`

With parameters [image 4a0-favicon-A.svg - 30%] and in this case scale do not work.
With no parameters doesn't work too — I see the 100% full page size scale image.

But if I use [image 4a0-favicon-A.svg - 300 300] it work well. 

So I think percent scale for percent sized images do not work. 
Open question about using the viewbox parameter in that case.

I have no time for deep explore before, so does this information can help?

**Comment by markseuffert** [comment, marked as answer]
Thanks for your feedback guys.

I am sorry to say, we don't support percentage scale for percentage sized images. You can edit your SVG so that they have a numerical width and height, e.g. `<svg width="256" height="256" ...>`. Then there's no guess work for the software, what the original image dimensions should be. Hope this helps.

**Comment by markseuffert** [comment]
I fixed the second problem that @GiovanniSalmeri found, please use the latest core. 

![Screen Shot 2019-03-13 at 10 07 33](https://user-images.githubusercontent.com/2429165/54266574-f24bf600-4577-11e9-964b-e24172bfd1f1.png)

You were right Giovanni, the regular expression was wrong. Good that you found it. About pull requests, you can pick any repository you like. I would recommend yellow-developers, then it's easy to test everything. Thanks again for your help.

**Comment by GiovanniSalmeri** [comment]
@markseu thank you very much!

After reading the specs of SVG I played a bit writing a version of `detectImageInformation` that recognises all the possible units (em, ex, px, in, cm, mm, pt, pc [with the usual assumptions about DPI and fonts]) and percentages. Honestly I do not know how many SVGs in the wild use these units... Is it useful if I post this version here, just in case?

On a sidenote: `detectImageInformation` dos not use the PHP function `getimagesize` because it would add a dependence on the GD extension of PHP, right?

**Comment by markseuffert** [comment]
Thanks for the offer. I personally think we should not support everything that's technical possible. Product design is partially about good technology, partially about finding what's good enough. Whenever a simple solution handles 90% of the uses cases and another solution that handles 100% has  the cost of additional complexity, then it's worth to consider which to pick. Sometimes less is more. 

Yes and I think it's faster. 😉

**Comment by GiovanniSalmeri** [comment]
Well, I agree that less is more, I have chosen Yellow exactly for this! :smile:

**Comment by i-do** [comment]
[Mark](@markseu), absolutely agree with you! Less is more.

But I think you can add little remark about svg scale, maybe add some samples of usage to manual.
